_posts/2018-12-07-fast-starting-clojure-lambdas-using-graalvm.md

```diff
@@ -0,0 +1,80 @@
+
+I made a simple custom runtime for running a Clojure program in a custom runtime. Compiling it was quite simple but required using a Docker container because binary must be compiled in Linux environment. Compilation times are  long compared to JVM compilation which must be made before compiling as a native image. The results are described in the next table. 
+
+| Memory | Runtime                     | N    | Average | Standard deviation |
+| ------ | --------------------------- | ---- | ------- | ------------------ |          
+| 1000   | GraalVM Clojure (cold)      | 1000 | 0.624   | 0.202              |
+| 1000   | GraalVM Clojure (warm)      | 5000 | 0.202   | 0.068              |
+| 1000   | JVM  Java (cold)            | 1000 | 0.585   | 0.108              |
+| 1000   | JVM  Java (warm)            | 5000 | 0.205   | 0.063              |
+| 1000   | JVM  Clojure (cold)         | 922  | 2.942   | 0.391              |
+| 1000   | JVM  Clojure (warm)         | 4600 | 0.237   | 0.133              |
+
+GraalVM makes Clojure run excellently in the Lambda environment. The cold start time is comparable to regular JVM. A quite interesting result is the warmed performance which is better in GraalVM than in the regular JVM.  
+
+## Limitations of GraalVM
+
+GraalVM have currently some problems compiling native images. For example no instances are allowed in the image heap for a class that is initialized or reinitialized at image runtime. These classes must be given as parameters which is cumbersome. The test program contained SSL libraries which caused compilation problems. Also certain libraries cannot be currently compiled. The Apache HTTP client which is used for example by Clojure Clj-http library is one of them. The compilation problem seems not to be Clojure specific so this should be fixed in the future versions of GraalVM. GraalVM added HTTPS protocol support to native-image in version 1.0.0-rc7, but it still has limitations. First, the provided certificate store has only limited set of CA certificates and second, you must configure path to libsunec.so (Sun Elliptic Curve crypto library). GraalVM tries to load the library from the current directory or from **java.library.path* when its first used. You can workaround these limitations by:
+
+1. Copy or make a symbolic link to the certificate store from e.g. your distributions OpenJDK to your GraalVM-installation. Certificate store is usually located in the file *$JDK_HOME/jre/lib/security/cacerts*.
+2. Configure path *java.library.path* to include the library *libsunec.so* (in Linux this is in directory *$GRAALVM_HOME/jre/lib/amd64/*) or copy the library file to the working directory).
+
+Runtime performance of the native images are [slightly worse](https://www.graalvm.org/docs/reference-manual/aot-compilation/) than regular JVM HotSpot compiler. This may of course change in the future. GraalVM is still in a release candidate phase for 1.0 version so situation may change in the future. 
```

Comment:
```suggestion
The runtime performance of the native images is [slightly worse](https://www.graalvm.org/docs/reference-manual/aot-compilation/) than regular JVM HotSpot compiler. This may of course change in the future. GraalVM is still in a release candidate phase for 1.0 version so the situation may change in the future. 
```